multidimensional-pod-autoscaler/pkg/updater/logic/updater.go

```diff
@@ -0,0 +1,524 @@
+			withEvictable = true
+			if !evictionLimiter.CanEvict(pod) {
+				continue
+			}
+			err := u.evictionRateLimiter.Wait(ctx)
+			if err != nil {
+				klog.Warningf("evicting pod %v failed: %v", pod.Name, err)
+				return
+			}
+			klog.V(2).Infof("evicting pod %v", pod.Name)
+			evictErr := evictionLimiter.Evict(pod, u.eventRecorder)
+			if evictErr != nil {
+				klog.Warningf("evicting pod %v failed: %v", pod.Name, evictErr)
+			} else {
+				withEvicted = true
+				metrics_updater.AddEvictedPod(mpaSize)
+			}
+		}
+
+		if withEvictable {
+			mpasWithEvictablePodsCounter.Add(mpaSize, 1)
+		}
+		if withEvicted {
+			mpasWithEvictedPodsCounter.Add(mpaSize, 1)
+		}
+	}
+	timer.ObserveStep("EvictPods")
+	klog.V(4).Infof("Evicted all eligible pods.")
+}
+
+// RunOnceUpdatingDeployment represents single iteration in the main-loop of Updater which evicts
+// pods for VPA and updates the Deployment for HPA.
+func (u *updater) RunOnceUpdatingDeployment(ctx context.Context) {
+	timer := metrics_updater.NewExecutionTimer()
+	defer timer.ObserveTotal()
+
+	if u.useAdmissionControllerStatus {
+		isValid, err := u.statusValidator.IsStatusValid(status.AdmissionControllerStatusTimeout)
+		if err != nil {
+			klog.Errorf("Error getting Admission Controller status: %v. Skipping eviction loop", err)
+			return
+		}
+		if !isValid {
+			klog.Warningf("Admission Controller status has been refreshed more than %v ago. Skipping eviction loop",
+				status.AdmissionControllerStatusTimeout)
+			return
+		}
+	}
+
+	mpaList, err := u.mpaLister.List(labels.Everything())
+	if err != nil {
+		klog.Fatalf("failed get MPA list: %v", err)
+	}
+	timer.ObserveStep("ListMPAs")
+	klog.V(4).Infof("Retrieved all MPA objects.")
+
+	mpas := make([]*mpa_api_util.MpaWithSelector, 0)
+
+	for _, mpa := range mpaList {
+		if mpa_api_util.GetUpdateMode(mpa) != vpa_types.UpdateModeRecreate &&
+			mpa_api_util.GetUpdateMode(mpa) != vpa_types.UpdateModeAuto {
+			klog.V(3).Infof("skipping MPA object %v because its mode is not \"Recreate\" or \"Auto\"", mpa.Name)
+			continue
+		}
+		selector, err := u.selectorFetcher.Fetch(mpa)
+		if err != nil {
+			klog.V(3).Infof("skipping MPA object %v because we cannot fetch selector", mpa.Name)
+			continue
+		}
+
+		mpas = append(mpas, &mpa_api_util.MpaWithSelector{
+			Mpa:      mpa,
+			Selector: selector,
+		})
+
+		// Update the number of replicas.
+		targetGV, err := schema.ParseGroupVersion(mpa.Spec.ScaleTargetRef.APIVersion)
+		if err != nil {
+			klog.Errorf("%s: FailedGetScale - error: %v", v1.EventTypeWarning, err.Error())
+			u.eventRecorder.Event(mpa, v1.EventTypeWarning, "FailedGetScale", err.Error())
+			return
+		}
+		fmt.Printf("targetGV %v targetGV.Group: %v\n", targetGV, targetGV.Group)
+		targetGK := schema.GroupKind{
+			Group: targetGV.Group,
+			Kind:  mpa.Spec.ScaleTargetRef.Kind,
+		}
+		fmt.Printf("targetGK: %v\n", targetGK)
+		mappings, err := u.selectorFetcher.GetRESTMappings(targetGK)
+		fmt.Printf("mapping: %v\n", mappings)
```

Comment:
I'm unsure if this sort of small comment is very useful right now, but since I'm looking at the code:
```suggestion
		mappings, err := u.selectorFetcher.GetRESTMappings(targetGK)
```